snap/snapdir/snapdir.go

```diff
@@ -60,6 +61,96 @@ func (s *SnapDir) ReadFile(file string) (content []byte, err error) {
 	return ioutil.ReadFile(filepath.Join(s.path, file))
 }
 
+// Walk (part of snap.Container) is like filepath.Walk, without the ordering guarantee.
+func (s *SnapDir) Walk(relative string, walkFn filepath.WalkFunc) error {
+	relative = filepath.Clean(relative)
+	if relative == "" || relative == "/" {
+		relative = "."
+	} else if relative[0] == '/' {
+		// I said relative, darn it :-)
+		relative = relative[1:]
+	}
+	root := filepath.Join(s.path, relative)
+	// we could just filepath.Walk(root, walkFn), but that doesn't scale
+	// well to insanely big directories as it reads the whole directory,
+	// in order to sort it. This Walk doesn't do that.
+	//
+	// Also the directory is always relative to the top of the container
+	// for us, which would make it a little more messy to get right.
+	f, err := os.Open(root)
+	if err != nil {
+		return walkFn(relative, nil, err)
+	}
+	defer func() {
+		if f != nil {
+			f.Close()
+		}
+	}()
+
+	st, err := f.Stat()
+	if err != nil {
+		return walkFn(relative, nil, err)
+	}
+
+	err = walkFn(relative, st, nil)
+	if err != nil {
+		return err
+	}
+	if !st.IsDir() {
+		return nil
+	}
+
+	const numSt = 100
+	var dirstack []string
+	for {
+		sts, err := f.Readdir(numSt)
```

Comment:
This long {} was bothering me a bit and I started to play around with things. What do you think about extracting a small helper? Like http://paste.ubuntu.com/26346230/ - to me it seems slightly easier to read/reason about, what do you think?

Reply:
well, `if len(dirstack)==0` bit would be broken with your patch, but it's fixable